lib/expand/join.go

```diff
@@ -504,6 +507,29 @@ func (p *Peer) getAgent(opCtx operationContext) (*rpcserver.PeerServer, error) {
 
 func (p *Peer) run() error {
 	ctx, err := p.connect()
+
+	// schedule a cleanup function to fail the operation if this exits
+	// with error
+	defer func() {
+		if err == nil {
+			return
+		}
+		p.WithError(err).Warn("Peer is exiting with error.")
+		stopCtx, cancel := context.WithTimeout(context.Background(), defaults.AgentStopTimeout)
+		defer cancel()
+		if p.agent != nil {
+			p.Info("Stopping peer.")
+			if err := p.agent.Stop(stopCtx); err != nil {
+				p.WithError(err).Error("Failed to stop peer.")
+			}
+		}
+		if ctx != nil && ctx.Operation.ID != "" {
```

Comment:
Should this also be executed during install operation, or should we check specifically for expand here?

Reply:
I don't see a reason to exclude install here.